syntax = "proto3";

import "std_buffer/image_metadata.proto";

package std_daq_protocol;

message CreateFileCommand {
  string path = 1;
  int32 writer_id = 2;
}

message RecordImage {
  ImageMetadata image_metadata = 1;
}

message CloseFile {
}

message ConfirmLastImageWritten {
}

message WriterAction {
  oneof action {
    CreateFileCommand create_file = 1;
    RecordImage record_image = 2;
    CloseFile close_file = 3;
    ConfirmLastImageWritten confirm_last_image = 4;
  }
}

enum CommandType {
  WRITE_IMAGE = 0;
  START_WRITING = 1;
  STOP_WRITING = 2;
}

message RunInfo {
  uint64 run_id = 1;
  string output_file = 2;
  int32 n_images = 3;
}

message WriterCommand {
  CommandType command_type = 1;
  RunInfo run_info = 2;
  uint32 i_image = 3;

  ImageMetadata metadata = 4;
}

message WriterStatus {
  CommandType command_type = 1;
  RunInfo run_info = 2;
  uint32 i_image = 3;

  uint32 i_writer = 4;
  uint32 n_writers = 5;
  string error_message = 6;
}
